packages/wallets/wallet-turnkey/src/strategy/strategy.ts

```diff
@@ -0,0 +1,732 @@
+/* eslint-disable class-methods-use-this */
+import {
+  ErrorType,
+  WalletException,
+  UnspecifiedErrorCode,
+  TransactionException,
+  CosmosWalletException,
+} from '@injectivelabs/exceptions'
+import {
+  TxRaw,
+  TxGrpcApi,
+  DirectSignResponse,
+  AminoSignResponse,
+  getInjectiveAddress,
+} from '@injectivelabs/sdk-ts'
+import { AccountAddress, EthereumChainId } from '@injectivelabs/ts-types'
+import {
+  StdSignDoc,
+  WalletAction,
+  WalletDeviceType,
+  BaseConcreteStrategy,
+  BrowserEip1993Provider,
+  ConcreteWalletStrategy,
+  SendTransactionOptions,
+  ConcreteEthereumWalletStrategyArgs,
+} from '@injectivelabs/wallet-base'
+import { SessionType, Turnkey, TurnkeyIframeClient } from '@turnkey/sdk-browser'
+
+import { createAccount } from '@turnkey/viem'
+
+import { TurnkeySDKBrowserConfig } from '@turnkey/sdk-browser'
+
+export type TurnkeyStatus =
+  | 'initializing'
+  | 'ready'
+  | 'waiting-otp'
+  | 'logged-in'
+  | 'error'
+
+export const TURNKEY_TOKEN_KEYS = {
+  TOKEN: 'turnkey-token',
+  ORGANIZATION_ID: 'turnkey-organization-id',
+  TOKEN_EXPIRATION: 'turnkey-token-expiration',
+} as const
+
+export type TurnkeyTokenValue =
+  (typeof TURNKEY_TOKEN_KEYS)[keyof typeof TURNKEY_TOKEN_KEYS]
+
+export type TurnkeyMetadata = TurnkeySDKBrowserConfig & {
+  turnkeyAuthIframeContainerId: string
+  turnkeyAuthIframeElementId?: string
+  token?: string
+  organizationId?: string
+}
+
+type OAuthArgs = {
+  provider: 'google'
+  oidcToken: string
+  oauthLoginEndpoint: string
+}
+
+type EmailArgs = {
+  provider: 'email'
+  email: string
+  initEmailOTPEndpoint: string
+}
+const refreshInterval_ms = 5 * 1000
+
+type TurnkeyEnableArgs = OAuthArgs | EmailArgs
+
+interface TurnkeyArgs extends ConcreteEthereumWalletStrategyArgs {
+  metadata?: TurnkeyMetadata
+  onStatusChange?: (status: TurnkeyStatus) => void
+}
+
+export class TurnkeyWallet
+  extends BaseConcreteStrategy
+  implements ConcreteWalletStrategy
+{
+  public provider: BrowserEip1993Provider | undefined
+  public metadata?: TurnkeyMetadata
+  public organizationId?: string
+  public turnkey: Turnkey
+  public authIframeClient?: TurnkeyIframeClient
+  public status: TurnkeyStatus = 'initializing'
+  public expiry?: number
+  private emailOTPId?: string
+  private onStatusChangeCallback?: (status: TurnkeyStatus) => void
+  private addresses: string[] = []
+
+  private accountAddresses: string[] = []
+  private accountMap: Record<
+    AccountAddress,
+    Awaited<ReturnType<typeof createAccount>>
+  > = {}
+
+  private setStatus(status: TurnkeyStatus) {
+    this.status = status
+    this.onStatusChangeCallback?.(status)
+  }
+
+  constructor(args: TurnkeyArgs) {
+    if (!args.metadata?.apiBaseUrl) {
```

Comment:
since we have more and more strategies using metadata, maybe we should name them properly i.e `turnKeyMetadata`, `walletConnectMetadata`, `privateKeyMetadata`, etc. 

cc @ThomasRalee

Reply:
Took a quick look, I like the flexibility on using metadata as the naming convention.
We will end up with n+1 arguments for the class, dont want to overbloat this if we can

Reply:
I've implemented this on the `injective-ts` level